File: agent/agent-harness/agent-harness/harness/api/orchestrator.py
"""Unified Scientific Workflow Orchestrator API routes (prompt §8.1). Every
route calls `harness.orchestrator.service.UnifiedScientificWorkflowOrchestrator`
- no business logic lives here, matching every other Problem's API router
convention in this repo (`harness/api/virtual_cell.py` etc).

Error mapping follows the same convention as the rest of this repo's API
(see README's API table): illegal phase transitions -> 409, stale-version
conflicts -> 409, gate/precondition rejections -> 422.
"""
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session

from harness.api.deps import get_db_session
from harness.db import ConcurrencyConflictError
from harness.designs.service import SelfApprovalError
from harness.diagnosis.loop import DiagnosisGateRejectedError, IllegalDiagnosisTransitionError
from harness.engineering_design.handoff import HandoffRejectedError
from harness.engineering_design.loop import DesignGateRejectedError, IllegalDesignTransitionError
from harness.engineering_design.portfolio_service import PortfolioDiversityRejected
from harness.engineering_design.project_service import ObjectiveRejected
from harness.orchestrator import service as orch_service
from harness.orchestrator.models import ModuleHandoffRecord, OrchestratorGateDecision, OrchestratorTransition, UnifiedWorkflowRun
from harness.scientific_evaluation.human_gate import HumanGatePreconditionError
from harness.scientific_evaluation.loop import EvaluationGateRejectedError, IllegalEvaluationTransitionError
from harness.workflow.iterative_loop import GateRejectedError, IllegalCycleTransitionError

router = APIRouter(prefix="/api/orchestrator", tags=["orchestrator"])
ORC = orch_service.UnifiedScientificWorkflowOrchestrator()


def _run_dict(r: UnifiedWorkflowRun) -> dict[str, Any]:
    return {
        "workflow_run_id": r.workflow_run_id, "project_id": r.project_id, "objective_id": r.objective_id,
        "dbtl_iteration_id": r.dbtl_iteration_id, "status": r.status, "current_phase": r.current_phase,
        "current_module": r.current_module, "diagnosis_run_ref": r.diagnosis_run_ref,
        "diagnosis_handoff_ref": r.diagnosis_handoff_ref, "design_project_ref": r.design_project_ref,
        "design_version_ref": r.design_version_ref, "evaluation_run_ref": r.evaluation_run_ref,
        "simulation_campaign_ref": r.simulation_campaign_ref, "experiment_plan_ref": r.experiment_plan_ref,
        "experiment_run_ref": r.experiment_run_ref, "observation_set_ref": r.observation_set_ref,
        "active_gate_ref": r.active_gate_ref, "pause_reason": r.pause_reason, "blocked_reason": r.blocked_reason,
        "checkpoint_ref": r.checkpoint_ref, "correlation_id": r.correlation_id, "created_at": r.created_at,
        "updated_at": r.updated_at, "version": r.version,
    }


def _handle(fn, *args, **kwargs):
    """Every orchestrator route drives a whole underlying module's loop
    controller/service layer (`DiagnosisAdapter`, `DesignAdapter`, etc,
    harness/orchestrator/adapters.py) - any domain-specific rejection that
    module's OWN API routes already handle (harness/api/{diagnosis,
    engineering_design,scientific_evaluation,projects}.py) can just as
    easily bubble up through this parallel orchestrator entry point, which
    never caught them. Found via a real, reproducible 500: `POST .../design`
    with no primary_metrics raised `ObjectiveRejected` here uncaught.
    """
    try:
        return fn(*args, **kwargs)
    except ConcurrencyConflictError as e:
        raise HTTPException(409, str(e))
    except (
        orch_service.OrchestratorPhaseError,
        orch_service.CycleConflictError,
        IllegalDiagnosisTransitionError,
        IllegalDesignTransitionError,
        IllegalEvaluationTransitionError,
        IllegalCycleTransitionError,
        SelfApprovalError,
    ) as e:
        raise HTTPException(409, str(e))
    except (
        ValueError,
        orch_service.OrchestratorBlockedError,
        DiagnosisGateRejectedError,
        DesignGateRejectedError,
        EvaluationGateRejectedError,
        GateRejectedError,
        ObjectiveRejected,
        PortfolioDiversityRejected,
        HumanGatePreconditionError,
        HandoffRejectedError,
    ) as e:
        raise HTTPException(422, str(e))


class CreateRunBody(BaseModel):
    project_id: str
    actor_id: str
    target_product: str
    host: str
    dbtl_iteration_id: str | None = None


@router.post("/runs")
def create_run(body: CreateRunBody, session: Session = Depends(get_db_session)) -> dict:
    run = _handle(ORC.create_run, session, project_id=body.project_id, actor_id=body.actor_id, target_product=body.target_product, host=body.host, dbtl_iteration_id=body.dbtl_iteration_id)
    return _run_dict(run)


@router.get("/runs")
def list_runs(project_id: str, session: Session = Depends(get_db_session)) -> dict:
    """Lists every `UnifiedWorkflowRun` for a project, most-recently-updated
    first - closes the "frontend cannot discover a project's run on its
    own" gap this module's own docstring flagged (`frontend/src/api/
    orchestrator.ts`): WorkspaceEntry/WorkspaceLayout use this to resume the
    latest real run instead of always minting a new empty one."""
    from sqlalchemy import select

    rows = session.execute(
        select(UnifiedWorkflowRun).where(UnifiedWorkflowRun.project_id == project_id).order_by(UnifiedWorkflowRun.updated_at.desc())
    ).scalars().all()
    return {"runs": [_run_dict(r) for r in rows]}


@router.get("/runs/{workflow_run_id}")
def get_run(workflow_run_id: str, session: Session = Depends(get_db_session)) -> dict:
    try:
        run = ORC.get_status(session, workflow_run_id)
    except ValueError as e:
        raise HTTPException(404, str(e))
    return _run_dict(run)


@router.get("/runs/{workflow_run_id}/reconcile")
def reconcile_run(workflow_run_id: str, session: Session = Depends(get_db_session)) -> dict:
    try:
        return ORC.reconcile(session, workflow_run_id)
    except ValueError as e:
        raise HTTPException(404, str(e))


@router.get("/runs/{workflow_run_id}/audit-trail")
def audit_trail(workflow_run_id: str, session: Session = Depends(get_db_session)) -> dict:
    from sqlalchemy import select

    transitions = session.execute(select(OrchestratorTransition).where(OrchestratorTransition.workflow_run_id == workflow_run_id).order_by(OrchestratorTransition.created_at)).scalars().all()
    gates = session.execute(select(OrchestratorGateDecision).where(OrchestratorGateDecision.workflow_run_id == workflow_run_id).order_by(OrchestratorGateDecision.timestamp)).scalars().all()
    return {
        "transitions": [
            {"transition_id": t.transition_id, "from_phase": t.from_phase, "to_phase": t.to_phase, "reason": t.reason, "actor_id": t.actor_id, "created_at": t.created_at}
            for t in transitions
        ],
        "gate_decisions": [
            {"gate_decision_id": g.gate_decision_id, "gate_type": g.gate_type, "decision": g.decision, "blocking_findings": g.blocking_findings,
             "non_blocking_findings": g.non_blocking_findings, "required_actions": g.required_actions, "actor": g.actor, "timestamp": g.timestamp}
            for g in gates
        ],
    }


@router.get("/runs/{workflow_run_id}/handoffs")
def handoffs(workflow_run_id: str, session: Session = Depends(get_db_session)) -> dict:
    from sqlalchemy import select

    rows = session.execute(select(ModuleHandoffRecord).where(ModuleHandoffRecord.workflow_run_id == workflow_run_id).order_by(ModuleHandoffRecord.created_at)).scalars().all()
    return {
        "handoffs": [
            {"handoff_id": h.handoff_id, "source_module": h.source_module, "target_module": h.target_module, "payload_refs": h.payload_refs,
             "unresolved_items": h.unresolved_items, "warnings": h.warnings, "confidence_status": h.confidence_status, "created_at": h.created_at}
            for h in rows
        ]
    }


class StartDiagnosisBody(BaseModel):
    expected_version: int
    actor_id: str
    request: dict[str, Any]
    context: dict[str, Any] = {}


@router.post("/runs/{workflow_run_id}/diagnosis")
def start_diagnosis(workflow_run_id: str, body: StartDiagnosisBody, session: Session = Depends(get_db_session)) -> dict:
    run = _handle(ORC.start_diagnosis, session, workflow_run_id, expected_version=body.expected_version, request=body.request, context=body.context, actor_id=body.actor_id)
    return _run_dict(run)


class ResumeDiagnosisBody(BaseModel):
    expected_version: int
    actor_id: str
    data_sufficiency: dict[str, bool]


@router.post("/runs/{workflow_run_id}/diagnosis/resume")
def resume_diagnosis(workflow_run_id: str, body: ResumeDiagnosisBody, session: Session = Depends(get_db_session)) -> dict:
    run = _handle(ORC.resume_diagnosis_with_data, session, workflow_run_id, expected_version=body.expected_version, data_sufficiency=body.data_sufficiency, actor_id=body.actor_id)
    return _run_dict(run)


class StartDesignBody(BaseModel):
    expected_version: int
    actor_id: str
    request: dict[str, Any]
    context: dict[str, Any] = {}


@router.post("/runs/{workflow_run_id}/design")
def start_design(workflow_run_id: str, body: StartDesignBody, session: Session = Depends(get_db_session)) -> dict:
    run = _handle(ORC.start_design, session, workflow_run_id, expected_version=body.expected_version, request=body.request, context=body.context, actor_id=body.actor_id)
    return _run_dict(run)


class EvaluateDesignPortfolioBody(BaseModel):
    expected_version: int
    actor_id: str


@router.post("/runs/{workflow_run_id}/design/evaluate-portfolio")
def evaluate_design_portfolio(workflow_run_id: str, body: EvaluateDesignPortfolioBody, session: Session = Depends(get_db_session)) -> dict:
    """Problem 4's own portfolio evaluation - a precondition for the
    build-governance sequence inside `/human-gate-decision` that
    `DesignAdapter.start()`'s auto-chain never triggered on its own.
    """
    run = _handle(ORC.evaluate_design_portfolio, session, workflow_run_id, expected_version=body.expected_version, actor_id=body.actor_id)
    return _run_dict(run)


class RunEvaluationBody(BaseModel):
    expected_version: int
    actor_id: str
    revision_limit: int = 3
    enable_llm_critic: bool = False


@router.post("/runs/{workflow_run_id}/evaluation")
def run_evaluation(workflow_run_id: str, body: RunEvaluationBody, session: Session = Depends(get_db_session)) -> dict:
    run = _handle(ORC.run_evaluation, session, workflow_run_id, expected_version=body.expected_version, actor_id=body.actor_id, revision_limit=body.revision_limit, enable_llm_critic=body.enable_llm_critic)
    return _run_dict(run)


class EvaluationHumanDecisionBody(BaseModel):
    expected_version: int
    decision: str  # one of harness.scientific_evaluation.models.HUMAN_DECISIONS
    actor_id: str
    approver_role: str = ""
    selected_candidates: list[str] = []
    rationale: str = ""
    revision_limit: int = 3


@router.post("/runs/{workflow_run_id}/evaluation/human-decision")
def record_evaluation_human_decision(workflow_run_id: str, body: EvaluationHumanDecisionBody, session: Session = Depends(get_db_session)) -> dict:
    """Records the scientific_evaluation module's own human decision AND
    re-gates the case so `current_phase` actually advances - unlike
    `POST /api/scientific-evaluation/evaluations/{id}/human-decision`,
    which only updates the EvaluationCase row and leaves the orchestrator's
    run stuck at phase=EVALUATION regardless of the decision.
    """
    run = _handle(
        ORC.record_evaluation_human_decision, session, workflow_run_id, expected_version=body.expected_version,
        decision=body.decision, actor_id=body.actor_id, approver_role=body.approver_role,
        selected_candidates=body.selected_candidates or None, rationale=body.rationale, revision_limit=body.revision_limit,
    )
    return _run_dict(run)


class SubmitRevisionBody(BaseModel):
    expected_version: int
    actor_id: str
    design_id: str
    modification_reason: str
    revision_limit: int = 3
    genetic_modifications: list[dict[str, Any]] | None = None


@router.post("/runs/{workflow_run_id}/evaluation/revise")
def submit_evaluation_revision(workflow_run_id: str, body: SubmitRevisionBody, session: Session = Depends(get_db_session)) -> dict:
    kwargs = {}
    if body.genetic_modifications is not None:
        kwargs["genetic_modifications"] = body.genetic_modifications
    run = _handle(
        ORC.submit_evaluation_revision, session, workflow_run_id, expected_version=body.expected_version, design_id=body.design_id,
        modification_reason=body.modification_reason, actor_id=body.actor_id, revision_limit=body.revision_limit, **kwargs,
    )
    return _run_dict(run)


class HumanGateDecisionBody(BaseModel):
    expected_version: int
    decision: str
    actor_id: str
    reason: str = ""
    selected_design_id: str | None = None
    build_test_kwargs: dict[str, Any] | None = None


@router.post("/runs/{workflow_run_id}/human-gate-decision")
def record_human_gate_decision(workflow_run_id: str, body: HumanGateDecisionBody, session: Session = Depends(get_db_session)) -> dict:
    run = _handle(
        ORC.record_human_gate_decision, session, workflow_run_id, expected_version=body.expected_version, decision=body.decision,
        actor_id=body.actor_id, reason=body.reason, selected_design_id=body.selected_design_id, build_test_kwargs=body.build_test_kwargs,
    )
    return _run_dict(run)


class RunSimulationBody(BaseModel):
    expected_version: int
    actor_id: str
    chassis: dict[str, Any] = {}
    environment: dict[str, Any] = {}
    model_id: str = "MREG-gem_fba"


@router.post("/runs/{workflow_run_id}/simulation")
def run_simulation(workflow_run_id: str, body: RunSimulationBody, session: Session = Depends(get_db_session)) -> dict:
    run = _handle(
        ORC.run_simulation, session, workflow_run_id, expected_version=body.expected_version, chassis=body.chassis,
        environment=body.environment, actor_id=body.actor_id, model_id=body.model_id,
    )
    return _run_dict(run)


class CreateExperimentPlanBody(BaseModel):
    expected_version: int
    actor_id: str
    hypotheses_tested: list[str] = []
    controls: list[str] = []
    factors: list[str] = []
    response_variables: list[str] = []
    acceptance_criteria: list[str] = []
    protocol_ref_id: str | None = None


@router.post("/runs/{workflow_run_id}/experiment-plan")
def create_experiment_plan(workflow_run_id: str, body: CreateExperimentPlanBody, session: Session = Depends(get_db_session)) -> dict:
    kwargs = body.model_dump(exclude={"expected_version", "actor_id"})
    run = _handle(ORC.create_experiment_plan, session, workflow_run_id, expected_version=body.expected_version, actor_id=body.actor_id, **kwargs)
    return _run_dict(run)


class RawObservationBody(BaseModel):
    feature_or_phenotype: str
    value: float | None = None
    unit: str | None = None
    condition_id: str | None = None
    qc_status: str | None = None
    replicates: int | None = None
    detection_limit: float | None = None
    assay_id: str | None = None
    timepoint: dict[str, Any] | None = None
    reference_or_baseline: dict[str, Any] | None = None


class RecordExperimentRunBody(BaseModel):
    expected_version: int
    actor_id: str
    raw_observation: RawObservationBody
    execution_status: str = "completed"
    deviations: list[str] = []
    operator_or_source: str = ""


@router.post("/runs/{workflow_run_id}/experiment-run")
def record_experiment_run(workflow_run_id: str, body: RecordExperimentRunBody, session: Session = Depends(get_db_session)) -> dict:
    from harness.diagnosis.normalizer import RawObservationInput

    raw = RawObservationInput(**body.raw_observation.model_dump())
    run_kwargs = {"execution_status": body.execution_status, "deviations": body.deviations, "operator_or_source": body.operator_or_source}
    run = _handle(
        ORC.record_experiment_run_and_ingest_observation, session, workflow_run_id, expected_version=body.expected_version,
        actor_id=body.actor_id, raw_observation=raw, run_kwargs=run_kwargs,
    )
    return _run_dict(run)


class RunLearningBody(BaseModel):
    expected_version: int
    actor_id: str
    observed_results: list[dict[str, Any]]
    construction_verified: bool
    assay_qc_passed: bool


@router.post("/runs/{workflow_run_id}/learning")
def run_learning(workflow_run_id: str, body: RunLearningBody, session: Session = Depends(get_db_session)) -> dict:
    run = _handle(
        ORC.run_learning, session, workflow_run_id, expected_version=body.expected_version, actor_id=body.actor_id,
        observed_results=body.observed_results, construction_verified=body.construction_verified, assay_qc_passed=body.assay_qc_passed,
    )
    return _run_dict(run)

File: frontend/src/api/orchestrator.ts
import { api } from "./client";

/**
 * Unified Scientific Workflow Orchestrator adapter (harness/api/orchestrator.py).
 * `DiagnosisAdapter.start()`/`DesignAdapter.start()` (harness/orchestrator/
 * adapters.py) already auto-chain diagnosis (session -> hypotheses ->
 * decision) and design (handoff -> objectives -> strategies -> portfolio)
 * in one call each - this client exposes exactly those two calls plus
 * `createRun`, so a project can go from "just created" to "has a
 * candidate portfolio" without a human clicking through every stage of
 * the manual Diagnosis/Engineering-Design workbenches. Evaluation/
 * build/human-approval stay manual on purpose - those are the real,
 * required governance gates (see `record_human_gate_decision`), not
 * unwired automation.
 */
export interface WorkflowRun {
  workflowRunId: string;
  projectId: string;
  objectiveId: string | null;
  dbtlIterationId: string | null;
  status: string;
  currentPhase: string;
  currentModule: string | null;
  diagnosisRunRef: string | null;
  diagnosisHandoffRef: string | null;
  designProjectRef: string | null;
  designVersionRef: string | null;
  evaluationRunRef: string | null;
  simulationCampaignRef: string | null;
  experimentPlanRef: string | null;
  experimentRunRef: string | null;
  observationSetRef: string[];
  activeGateRef: string | null;
  pauseReason: string | null;
  blockedReason: string | null;
  checkpointRef: string | null;
  correlationId: string | null;
  createdAt: number;
  updatedAt: number;
  version: number;
}

interface RawRun {
  workflow_run_id: string; project_id: string; objective_id: string | null; dbtl_iteration_id: string | null;
  status: string; current_phase: string; current_module: string | null; diagnosis_run_ref: string | null;
  diagnosis_handoff_ref: string | null; design_project_ref: string | null; design_version_ref: string | null;
  evaluation_run_ref: string | null; simulation_campaign_ref: string | null; experiment_plan_ref: string | null;
  experiment_run_ref: string | null; observation_set_ref: string[]; active_gate_ref: string | null;
  pause_reason: string | null; blocked_reason: string | null; checkpoint_ref: string | null;
  correlation_id: string | null; created_at: number; updated_at: number; version: number;
}

function toRun(r: RawRun): WorkflowRun {
  return {
    workflowRunId: r.workflow_run_id, projectId: r.project_id, objectiveId: r.objective_id,
    dbtlIterationId: r.dbtl_iteration_id, status: r.status, currentPhase: r.current_phase,
    currentModule: r.current_module, diagnosisRunRef: r.diagnosis_run_ref, diagnosisHandoffRef: r.diagnosis_handoff_ref,
    designProjectRef: r.design_project_ref, designVersionRef: r.design_version_ref, evaluationRunRef: r.evaluation_run_ref,
    simulationCampaignRef: r.simulation_campaign_ref, experimentPlanRef: r.experiment_plan_ref,
    experimentRunRef: r.experiment_run_ref, observationSetRef: r.observation_set_ref, activeGateRef: r.active_gate_ref,
    pauseReason: r.pause_reason, blockedReason: r.blocked_reason, checkpointRef: r.checkpoint_ref,
    correlationId: r.correlation_id, createdAt: r.created_at, updatedAt: r.updated_at, version: r.version,
  };
}

/** `POST /api/orchestrator/runs`. */
export async function createRun(input: { projectId: string; actorId: string; targetProduct: string; host: string; dbtlIterationId?: string }): Promise<WorkflowRun> {
  const r = await api.post<RawRun>("/api/orchestrator/runs", {
    project_id: input.projectId, actor_id: input.actorId, target_product: input.targetProduct,
    host: input.host, dbtl_iteration_id: input.dbtlIterationId ?? null,
  });
  return toRun(r);
}

/** `GET /api/orchestrator/runs?project_id=` - most-recently-updated first. */
export async function listRuns(projectId: string): Promise<WorkflowRun[]> {
  const raw = await api.get<{ runs: RawRun[] }>(`/api/orchestrator/runs?project_id=${encodeURIComponent(projectId)}`);
  return raw.runs.map(toRun);
}

/** `GET /api/orchestrator/runs/{id}`. */
export async function getRun(workflowRunId: string): Promise<WorkflowRun> {
  const r = await api.get<RawRun>(`/api/orchestrator/runs/${workflowRunId}`);
  return toRun(r);
}

export interface StartDiagnosisInput {
  expectedVersion: number;
  actorId: string;
  biologicalSystem: Record<string, unknown>;
  phenotype: string;
  targetProduct: string;
  host: string;
  observationIds: string[];
  baselineObservationIds: string[];
  dataSufficiency: {
    hasBaseline: boolean; hasGenotype: boolean; hasCondition: boolean;
    hasTime: boolean; hasQc: boolean; hasKeyPhenotype: boolean;
  };
  context?: Record<string, unknown>;
}

/** `POST /api/orchestrator/runs/{id}/diagnosis` - auto-chains session creation
 * through hypothesis generation to a `DiagnosisDecision` in one call
 * (`DiagnosisAdapter.start()`/`_run_hypothesis_pipeline`), stopping only at
 * the legitimate `data_required`/`human_review_required` checkpoints. */
export async function startDiagnosis(workflowRunId: string, input: StartDiagnosisInput): Promise<WorkflowRun> {
  const r = await api.post<RawRun>(`/api/orchestrator/runs/${workflowRunId}/diagnosis`, {
    expected_version: input.expectedVersion, actor_id: input.actorId,
    request: {
      biological_system: input.biologicalSystem, phenotype: input.phenotype, target_product: input.targetProduct,
      host: input.host,
      observation_ids: input.observationIds,
      baseline_observation_ids: input.baselineObservationIds,
      data_sufficiency: {
        has_baseline: input.dataSufficiency.hasBaseline, has_genotype: input.dataSufficiency.hasGenotype,
        has_condition: input.dataSufficiency.hasCondition, has_time: input.dataSufficiency.hasTime,
        has_qc: input.dataSufficiency.hasQc, has_key_phenotype: input.dataSufficiency.hasKeyPhenotype,
      },
    },
    context: input.context ?? {},
  });
  return toRun(r);
}

export interface StartDesignInput {
  expectedVersion: number;
  actorId: string;
  chassis: string;
  chassisVersionOrGenotype: string;
  primaryMetrics: unknown[];
  hardConstraints: unknown[];
  availableResources?: Record<string, unknown>;
  context?: Record<string, unknown>;
}

/** `POST /api/orchestrator/runs/{id}/design` - auto-chains handoff ingestion
 * through objective confirmation, strategy generation, and portfolio
 * generation in one call (`DesignAdapter.start()`). */
export async function startDesign(workflowRunId: string, input: StartDesignInput): Promise<WorkflowRun> {
  const r = await api.post<RawRun>(`/api/orchestrator/runs/${workflowRunId}/design`, {
    expected_version: input.expectedVersion, actor_id: input.actorId,
    request: {
      chassis: input.chassis, chassis_version_or_genotype: input.chassisVersionOrGenotype,
      primary_metrics: input.primaryMetrics, hard_constraints: input.hardConstraints,
      available_resources: input.availableResources ?? {},
    },
    context: input.context ?? {},
  });
  return toRun(r);
}

/** `POST /api/orchestrator/runs/{id}/design/evaluate-portfolio`. */
export async function evaluateDesignPortfolio(workflowRunId: string, expectedVersion: number, actorId: string): Promise<WorkflowRun> {
  const r = await api.post<RawRun>(`/api/orchestrator/runs/${workflowRunId}/design/evaluate-portfolio`, {
    expected_version: expectedVersion, actor_id: actorId,
  });
  return toRun(r);
}

export interface OrchestratorTransitionRow {
  transitionId: string;
  fromPhase: string;
  toPhase: string;
  reason: string;
  actorId: string;
  createdAt: number;
}

/** `GET /api/orchestrator/runs/{id}/audit-trail` - transitions only (gate decisions omitted; not needed by this UI yet). */
export async function getAuditTrail(workflowRunId: string): Promise<OrchestratorTransitionRow[]> {
  const raw = await api.get<{ transitions: Array<{ transition_id: string; from_phase: string; to_phase: string; reason: string; actor_id: string; created_at: number }> }>(
    `/api/orchestrator/runs/${workflowRunId}/audit-trail`,
  );
  return raw.transitions.map((t) => ({ transitionId: t.transition_id, fromPhase: t.from_phase, toPhase: t.to_phase, reason: t.reason, actorId: t.actor_id, createdAt: t.created_at }));
}
